### PR TITLE
Add support for documenting multiple instances of a method

### LINKED ISSUE #451: assert_rst_strings_are_equal() can result false positives
The function completely ignores blank lines.
In reality, in an RST string a single blank line is equal to 42 contiguous blank lines. But it's not the same as zero blank lines.

How to fix:

a) splitlines
b) collapse multiple contiguous blank lines into one
c) assert len() is equal
d) strip trailing whitespace in each element
e) assert each line is equal

This improves further on #448

### PR DESCRIPTION
**Pull request by texasaggie97**
[X] This contribution adheres to [CONTRIBUTING.md](https://github.com/ni/nimi-python/blob/master/CONTRIBUTING.md).
~~[ ] I've updated [CHANGELOG.md](https://github.com/ni/nimi-python/blob/master/CHANGELOG.md) if applicable.~~
~~[ ] I've added tests applicable for this pull request~~
### What does this Pull Request accomplish?
* We can now generate multiple method documentation for a given function:
  * Normal Python types
  * High performance `numpy` types
* We need to generate one entry per public method that we generate
  * Use the same mechanism of multiple templates to generate the different entries
* assert_rst_strings_are_equal() now correctly collapses multiple empty lines

### List issues fixed by this Pull Request below, if any.
* Fixes #451

### What testing has been done?
* Unit
* System
* Visual inspection of generated documentation

### CODE REVIEW

File: build/helper/documentation_helper.py [hunk 1]
```diff
@@ -296,15 +296,20 @@ def _fix_references(doc, cfg, make_link=False):
     return doc
 
 
-def _format_type_for_rst_documentation(param, config):
-    p_type = param['python_type']
-    if param['enum'] is not None:
+def format_type_for_rst_documentation(param, config, numpy):
```
Comment (marcoskirsch) [inline build/helper/documentation_helper.py]: Usually config goes at the end.
Applies to several places.

File: build/helper/documentation_helper.py [hunk 2]
```diff
@@ -505,14 +533,26 @@ def as_rest_table(data, header=True):
 
 def _remove_trailing_whitespace(s):
     '''Removes trailing whitespace and empty lines in multi-line strings. https://stackoverflow.com/a/17350806/316875'''
```
Comment (marcoskirsch) [inline build/helper/documentation_helper.py]: stale URL?

File: build/helper/documentation_helper.py [hunk 3]
```diff
@@ -505,14 +533,26 @@ def as_rest_table(data, header=True):
...
 
 
 def assert_rst_strings_are_equal(expected, actual):
     '''Asserts rst formatted strings (multiline) are equal. Ignores trailing whitespace and empty lines.'''
-    expected = _remove_trailing_whitespace(expected).splitlines()
-    actual = _remove_trailing_whitespace(actual).splitlines()
-    assert len(expected) == len(actual), 'Strings have different number of non-empty lines.'
+    expected = _remove_trailing_whitespace(expected)
+    actual = _remove_trailing_whitespace(actual)
+    # assert len(expected) == len(actual), 'Strings have different number of non-empty lines.'
```
Comment (marcoskirsch) [inline build/helper/documentation_helper.py]: remove

File: build/helper/documentation_helper.py [hunk 4]
```diff
@@ -522,6 +562,7 @@ def assert_rst_strings_are_equal(expected, actual):
         'GetTurtleID': {
             'codegen_method': 'public',
             'returns': 'ViStatus',
+            'method_templates': [{'filename': '/default_method', 'suffix': '', }, ],
```
Comment (marcoskirsch) [inline build/helper/documentation_helper.py]: In this context, one would think 'suffix' refers to the method template.
It actually refers to a suffix to be added to to the python_name for the method.

key should be renamed ('method_python_name_suffix'?), and plase add info about this to https://github.com/ni/nimi-python/wiki/API-Metadata

File: build/helper/documentation_helper.py (hunk 1 above)
Comment (texasaggie97) [reply build/helper/documentation_helper.py]: Moved

File: build/helper/documentation_helper.py (hunk 2 above)
Comment (texasaggie97) [reply build/helper/documentation_helper.py]: Removed

File: build/helper/documentation_helper.py (hunk 3 above)
Comment (texasaggie97) [reply build/helper/documentation_helper.py]: While asserting that actual and expected strings have the same number of lines is more correct, it actually makes it harder to debug differences.

Removed commented out assert

File: build/helper/documentation_helper.py (hunk 4 above)
Comment (texasaggie97) [reply build/helper/documentation_helper.py]: Changed code to use 'method_python_name_suffix'

I will add documentation for 'method_templates' to that page.

File: build/helper/metadata_add_all.py [hunk 5]
```diff
@@ -166,15 +166,13 @@ def _add_default_value_name_for_docs(parameter, module_name):
 _repeated_capability_parameter_names = ['channelName', 'channelList', 'channel', 'channelNameList']
 
 
-def _add_method_template_filenames(f):
+def _add_method_templates(f):
     '''Adds a list of 'method_template_filenames' value to function metadata if not found. This are the mako templates that will be used to render the method.'''
```
Comment (marcoskirsch) [inline build/helper/metadata_add_all.py]: Update docstring

File: generated/nifake/session.py [hunk 6]
```diff
@@ -560,9 +560,10 @@ def fetch_waveform_into(self, number_of_samples, waveform_data):
 
         Args:
             number_of_samples (int): Number of samples to return
+            waveform_data (numpy array of float64): Samples fetched from the device. Array should be numberOfSamples big.
 
         Returns:
-            waveform_data (list of float): Samples fetched from the device. Array should be numberOfSamples big.
+            waveform_data (numpy array of float64): Samples fetched from the device. Array should be numberOfSamples big.
```
Comment (marcoskirsch) [inline generated/nifake/session.py]: We don't really return anything, this is misleading.

File: generated/nifake/session.py [hunk 7]
```diff
@@ -1074,7 +1075,8 @@ def write_waveform_numpy(self, waveform):
         Writes waveform to the driver
 
         Args:
-            waveform (list of float): Waveform data.
+            number_of_samples (int): How many samples the waveform contains.
+            waveform (numpy array of float64): Waveform data.
```
Comment (marcoskirsch) [inline generated/nifake/session.py]: Maybe we can append something to this to clarify that the fetched data is copied onto this argument.

File: build/helper/metadata_add_all.py (hunk 5 above)
Comment (texasaggie97) [reply build/helper/metadata_add_all.py]: Updated

File: generated/nifake/session.py (hunk 6 above)
Comment (texasaggie97) [reply generated/nifake/session.py]: Should we also return it?

**Pull request merged**